<a href="https://colab.research.google.com/github/EdioHub/Score_Credito/blob/main/Score_cr%C3%A9dito.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Projeto Python IA: Inteligência Artificial e Previsões

### Case: Score de Crédito de Clientes  de um banco

 Análise de todos os clientes do banco para a criação de modelos que consigam ler as informações do cliente e dizer automaticamente o score de crédito dele: Ruim, Ok, Bom

 Uso de modelos comparativos entre KNN, Naives, SVC, Ada, Forest, Tree


In [ ]:
import pandas as pd


In [ ]:
import numpy as np
import sklearn

In [ ]:
data = pd.read_csv('/content/clientes.csv')
data.sample(5)

id_cliente  mes  idade   profissao  salario_anual  num_contas  \
3507       17350    4   25.0  jornalista      24195.370         7.0   
6049       31159    2   36.0  jornalista      40514.840         2.0   
461        40312    6   21.0   cientista      69977.600        10.0   
5623       30022    8   24.0  empresario      11866.345         7.0   
8322       26978    3   24.0   professor      40198.960         4.0   

      num_cartoes  juros_emprestimo  num_emprestimos  dias_atraso  ...  \
3507          7.0              19.0              1.0          8.0  ...   
6049          5.0               3.0              3.0          2.0  ...   
461           7.0              20.0              9.0         54.0  ...   
5623          7.0              20.0              1.0         17.0  ...   
8322          6.0              10.0              2.0          9.0  ...   

      idade_historico_credito  investimento_mensal  \
3507                    355.0            34.857220   
6049                    213.0            27.496983   
461                     157.0            64.662003   
5623                    296.0            15.249819   
8322                    335.0            37.326929   

          comportamento_pagamento  saldo_final_mes  score_credito  \
3507   baixo_gasto_pagamento_alto       329.252213       Standard   
6049  baixo_gasto_pagamento_medio       456.935753           Good   
461   baixo_gasto_pagamento_medio        59.886193           Poor   
5623   baixo_gasto_pagamento_alto       273.374856       Standard   
8322  baixo_gasto_pagamento_baixo       331.261865       Standard   

      emprestimo_carro  emprestimo_casa emprestimo_pessoal  \
3507               0.0              0.0                0.0   
6049               1.0              0.0                0.0   
461                1.0              1.0                1.0   
5623               0.0              0.0                0.0   
8322               0.0              0.0                0.0   

      emprestimo_credito emprestimo_estudantil  
3507                 1.0                   0.0  
6049                 0.0                   0.0  
461                  0.0                   1.0  
5623                 1.0                   0.0  
8322                 0.0                   0.0  

[5 rows x 25 columns]

In [ ]:
data['profissao'].unique()

array(['cientista', 'professor', 'engenheiro', 'empresario',
       'desenvolvedor', 'advogado', 'gerente_midia', 'medico',
       'jornalista', 'gerente', 'contador', 'musico', 'mecanico',
       'escritor', 'arquiteto'], dtype=object)

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11570 entries, 0 to 11569
Data columns (total 25 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id_cliente                11570 non-null  int64  
 1   mes                       11570 non-null  int64  
 2   idade                     11570 non-null  float64
 3   profissao                 11570 non-null  object 
 4   salario_anual             11570 non-null  float64
 5   num_contas                11570 non-null  float64
 6   num_cartoes               11569 non-null  float64
 7   juros_emprestimo          11569 non-null  float64
 8   num_emprestimos           11569 non-null  float64
 9   dias_atraso               11569 non-null  float64
 10  num_pagamentos_atrasados  11569 non-null  float64
 11  num_verificacoes_credito  11569 non-null  float64
 12  mix_credito               11569 non-null  object 
 13  divida_total              11569 non-null  float64
 14  taxa_u

In [ ]:
data.describe()

id_cliente           mes         idade  salario_anual    num_contas  \
count  11570.000000  11570.000000  11570.000000   11570.000000  11570.000000   
mean   26023.476923      4.499481     33.267589   51621.728545      5.328176   
std    14519.639701      2.291538     10.696803   38202.020335      2.650157   
min     1030.000000      1.000000     14.000000    7006.520000      0.000000   
25%    13651.000000      2.000000     25.000000   19981.600000      3.000000   
50%    25763.000000      4.000000     33.000000   39232.160000      6.000000   
75%    39249.000000      6.000000     42.000000   73434.920000      7.000000   
max    50984.000000      8.000000     56.000000  179948.840000     10.000000   

        num_cartoes  juros_emprestimo  num_emprestimos   dias_atraso  \
count  11569.000000      11569.000000     11569.000000  11569.000000   
mean       5.515602         14.327427         3.507390     21.174345   
std        2.040255          8.764199         2.444714     14.575065   
min        0.000000          1.000000         0.000000      0.000000   
25%        4.000000          7.000000         2.000000     10.000000   
50%        5.000000         13.000000         3.000000     18.000000   
75%        7.000000         20.000000         5.000000     28.000000   
max       11.000000         34.000000         9.000000     62.000000   

       num_pagamentos_atrasados  ...  divida_total  taxa_uso_credito  \
count              11569.000000  ...  11569.000000      11569.000000   
mean                  13.148846  ...   1405.194160         32.259241   
std                    6.296951  ...   1168.681338          5.153257   
min                    0.000000  ...      1.480000         20.172942   
25%                    9.000000  ...    526.620000         27.925878   
50%                   14.000000  ...   1131.700000         32.269861   
75%                   18.000000  ...   1873.810000         36.503076   
max                   25.000000  ...   4998.070000         50.000000   

       idade_historico_credito  investimento_mensal  saldo_final_mes  \
count             11569.000000         11569.000000     11569.000000   
mean                223.815369            55.668543       396.406342   
std                  98.052668            38.352132       204.380160   
min                   1.000000             0.000000         0.688299   
25%                 150.000000            28.094265       269.033254   
50%                 221.000000            46.753726       337.938190   
75%                 305.000000            72.438558       467.233460   
max                 404.000000           256.313483      1183.930696   

       emprestimo_carro  emprestimo_casa  emprestimo_pessoal  \
count      11569.000000     11569.000000        11569.000000   
mean           0.274527         0.494425            0.332613   
std            0.446295         0.499991            0.471170   
min            0.000000         0.000000            0.000000   
25%            0.000000         0.000000            0.000000   
50%            0.000000         0.000000            0.000000   
75%            1.000000         1.000000            1.000000   
max            1.000000         1.000000            1.000000   

       emprestimo_credito  emprestimo_estudantil  
count        11569.000000           11569.000000  
mean             0.316708               0.313942  
std              0.465213               0.464113  
min              0.000000               0.000000  
25%              0.000000               0.000000  
50%              0.000000               0.000000  
75%              1.000000               1.000000  
max              1.000000               1.000000  

[8 rows x 21 columns]

In [ ]:
from sklearn.preprocessing import LabelEncoder
codificador = LabelEncoder()

In [ ]:
data['profissao']=codificador.fit_transform(data['profissao'])
data['mix_credito']=codificador.fit_transform(data['mix_credito'])
data['comportamento_pagamento']=codificador.fit_transform(data['comportamento_pagamento'])


In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11570 entries, 0 to 11569
Data columns (total 25 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   id_cliente                11570 non-null  int64  
 1   mes                       11570 non-null  int64  
 2   idade                     11570 non-null  float64
 3   profissao                 11570 non-null  int64  
 4   salario_anual             11570 non-null  float64
 5   num_contas                11570 non-null  float64
 6   num_cartoes               11569 non-null  float64
 7   juros_emprestimo          11569 non-null  float64
 8   num_emprestimos           11569 non-null  float64
 9   dias_atraso               11569 non-null  float64
 10  num_pagamentos_atrasados  11569 non-null  float64
 11  num_verificacoes_credito  11569 non-null  float64
 12  mix_credito               11570 non-null  int64  
 13  divida_total              11569 non-null  float64
 14  taxa_u

In [ ]:
val = data.isna().any()
print(val)

id_cliente                  False
mes                         False
idade                       False
profissao                   False
salario_anual               False
num_contas                  False
num_cartoes                  True
juros_emprestimo             True
num_emprestimos              True
dias_atraso                  True
num_pagamentos_atrasados     True
num_verificacoes_credito     True
mix_credito                 False
divida_total                 True
taxa_uso_credito             True
idade_historico_credito      True
investimento_mensal          True
comportamento_pagamento     False
saldo_final_mes              True
score_credito                True
emprestimo_carro             True
emprestimo_casa              True
emprestimo_pessoal           True
emprestimo_credito           True
emprestimo_estudantil        True
dtype: bool


In [ ]:
linhas_com_ausentes = data[data.isna().any(axis=1)]
linhas_com_ausentes

id_cliente  mes  idade  profissao  salario_anual  num_contas  \
11569       50746    2   20.0          3      18438.875         0.0   

       num_cartoes  juros_emprestimo  num_emprestimos  dias_atraso  ...  \
11569          NaN               NaN              NaN          NaN  ...   

       idade_historico_credito  investimento_mensal  comportamento_pagamento  \
11569                      NaN                  NaN                        6   

       saldo_final_mes  score_credito  emprestimo_carro  emprestimo_casa  \
11569              NaN            NaN               NaN              NaN   

       emprestimo_pessoal  emprestimo_credito emprestimo_estudantil  
11569                 NaN                 NaN                   NaN  

[1 rows x 25 columns]

In [ ]:
data.dropna(inplace=True)
linhas_com_ausentes = data[data.isna().any(axis=1)]
linhas_com_ausentes

Empty DataFrame
Columns: [id_cliente, mes, idade, profissao, salario_anual, num_contas, num_cartoes, juros_emprestimo, num_emprestimos, dias_atraso, num_pagamentos_atrasados, num_verificacoes_credito, mix_credito, divida_total, taxa_uso_credito, idade_historico_credito, investimento_mensal, comportamento_pagamento, saldo_final_mes, score_credito, emprestimo_carro, emprestimo_casa, emprestimo_pessoal, emprestimo_credito, emprestimo_estudantil]
Index: []

[0 rows x 25 columns]

In [ ]:
y = data['score_credito']
x = data.drop(['score_credito','id_cliente'], axis = 1)

In [ ]:
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.3)

In [ ]:
from sklearn.ensemble import RandomForestClassifier, AdaBoostClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.gaussian_process import GaussianProcessClassifier
from sklearn.linear_model import Perceptron

In [ ]:
model_forest = RandomForestClassifier()
model_knn = KNeighborsClassifier()
model_naives = GaussianNB()
model_tree = DecisionTreeClassifier()
model_svc = SVC()
model_ada = AdaBoostClassifier()
model_gausian = GaussianProcessClassifier()
model_perceptron = Perceptron()

In [ ]:
model_forest.fit(x_train, y_train)


RandomForestClassifier()

In [ ]:
model_knn.fit(x_train, y_train)

KNeighborsClassifier()

In [ ]:
model_naives.fit(x_train, y_train)

GaussianNB()

In [ ]:
model_tree.fit(x_train, y_train)

DecisionTreeClassifier()

In [ ]:
model_ada.fit(x_train, y_train)

AdaBoostClassifier()

In [ ]:
model_svc.fit(x_train, y_train)

SVC()

In [ ]:
model_gausian.fit(x_train, y_train)

GaussianProcessClassifier()

In [ ]:
model_perceptron.fit(x_train, y_train)

Perceptron()

In [ ]:
from sklearn.metrics import accuracy_score

In [ ]:
previsao_forest = model_forest.predict(x_test)
previsao_knn = model_knn.predict(x_test)
previsao_naives = model_naives.predict(x_test)
previsao_tree = model_tree.predict(x_test)
previsao_ada = model_ada.predict(x_test)
previsao_svc = model_svc.predict(x_test)
previsao_gausian = model_gausian.predict(x_test)
previsao_perceptron = model_perceptron.predict(x_test)

In [ ]:
print('forest = ', accuracy_score(y_test,previsao_forest))
print('Knn = ', accuracy_score(y_test,previsao_knn))
print('naives = ', accuracy_score(y_test,previsao_naives))
print('tree = ', accuracy_score(y_test,previsao_tree))
print('ada = ', accuracy_score(y_test,previsao_ada))
print('svc = ', accuracy_score(y_test,previsao_svc))
print('gausian = ', accuracy_score(y_test,previsao_gausian))
print('perceptron = ', accuracy_score(y_test,previsao_perceptron) )

forest =  0.8352059925093633
Knn =  0.7853644482857965
naives =  0.592048401037165
tree =  0.7738403918179199
ada =  0.6502448861999424
svc =  0.5203111495246326
gausian =  0.5600691443388073
perceptron =  0.31892826274848746


# Conclusão

Com base nos resultados obtidos a partir da avaliação de diversos modelos de aprendizado de máquina, fica evidente que o Modelo Forest se destacou como a escolha mais eficaz para este sistema preditivo. Com uma taxa de acurácia de aproximadamente 83,52%, o Modelo Forest demonstrou sua capacidade de realizar previsões com alta precisão, tornando-se uma ferramenta valiosa para tomar decisões embasadas.

Comparativamente, outros modelos como K-Nearest Neighbors (Knn), Decision Tree, AdaBoost e Naive Bayes apresentaram taxas de acurácia inferiores, variando entre 78,53% e 31,89%. Isso ressalta a superioridade do Modelo Forest no contexto deste estudo específico.

No entanto, é importante ressaltar que o Support Vector Classifier (SVC) não obteve um desempenho satisfatório, atingindo uma taxa de acurácia de apenas 52,03% e o gausian com 56%. Além disso, observou-se que o SVC e o gausian demonstraram uma notável lentidão devido à complexidade de suas modelagens. Esse fator, aliado ao baixo desempenho, torna o SVC e o gausian inadequados para serem utilizados em sistemas que lidam com bancos de dados de grande porte.
o modelo de perceptron também não se mostrou efetivo para essa análise, obtendo a pior acurácia,31,8% ou seja mais erra do que acerta,provavelmente uma rede profunda seria muito mais efetiva, mas foge ao escopop dessa análise.

Diante desses resultados, conclui-se que a seleção do modelo de aprendizado de máquina é crucial para alcançar previsões precisas e eficientes. O Modelo Forest se sobressaiu, oferecendo um equilíbrio entre acurácia e desempenho, sendo altamente recomendado para a implementação neste sistema preditivo. Por outro lado, o uso do SVC deve ser evitado, especialmente em cenários que demandam rapidez e lidam com volumes significativos de dados. Este estudo reforça a importância de considerar não apenas a acurácia, mas também fatores como eficiência e escalabilidade ao escolher um modelo de aprendizado de máquina para tarefas preditivas.